### PR TITLE
Added functions to parse incoming MIME messages

### PR DESCRIPTION
**Pull request by yonigoogle**
Added functions to parse incoming MIME messages (based on Yan Zhu's original pull request)

Addressed a number of open issues in the existing library, including verifying that the boundary for a MIME message is unique.
Parsing an incoming message is done by calling parseNode(), a function that returns a treelike-object representing the underlying
MIME message.
Many of the functions are public, to give callers greater flexibility when using the library.

@diracdeltas @adhintz @kbsriram @koto @sirdarckcat

### CODE REVIEW

File: src/javascript/crypto/e2e/openpgp/pgpmime/constants.js [hunk 1]
```diff
@@ -56,10 +56,24 @@ e2e.openpgp.pgpmime.Constants.Mime = {
...
+  IMAGE_JPEG: 'image/jpeg',
+  IMAGE_JPG: 'image/jpg',
+  IMAGE_PNG: 'image/png',
+  IMAGE_BMP: 'image/bmp',
+  IMAGE_WEBP: 'image/webp',
+  IMAGE_PJPEG: 'image/pjpeg',
+  IMAGE_TIFF: 'image/tiff',
+  IMAGE_SVG_XML: 'image/svg+xml',
+  IMAGE_VND_DJVU: 'image/vnd.djvu',
+
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/constants.js]: Where did you get the list of image types from?

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 2]
```diff
@@ -130,10 +125,66 @@ e2e.openpgp.pgpmime.MimeNode.prototype.setContent = function(content) {
...
+/**
+ * Sets the MIME message boundary for a node.
+ * Until it succeeds, this function recursively reattempts to generate a unique
+ * boundary for the message (i.e., it attempts to generate a string that doesn't
+ * appear within the original message).
+ * @param {number} bound A bound on the amount of permitted recursions.
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setBoundary_ = function(bound) {
+  var boundary = this.generateBoundary_();
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: A `do..while` loop is cleaner here :

do {generate } while (!isValid);

I don't think there's a need for the recursion limit if the boundary has enough entropy and is unpredictable. To achieve that you can use e2e.random.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 3]
```diff
@@ -130,10 +125,66 @@ e2e.openpgp.pgpmime.MimeNode.prototype.setContent = function(content) {
...
+};
+
+
+/**
  * Builds an RFC 2822 message from the node and its children.
  * @return {string}
+ * @private
  */
-e2e.openpgp.pgpmime.MimeNode.prototype.buildMessage = function() {
+e2e.openpgp.pgpmime.MimeNode.prototype.mimeToString_ = function() {
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: You could just override toString().

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode_test.js [hunk 4]
```diff
@@ -184,3 +183,51 @@ function testBuildMessage() {
   var builtMessage = node.buildMessage();
   assertEquals(FINAL_MESSAGE, builtMessage);
 }
+
+function testisBoundaryValid() {
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode_test.js]: testIs...(capital I)

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 3 above)
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: IMO `buildMessage` is a better replacement for `toString`.

File: src/javascript/crypto/e2e/openpgp/pgpmime/utils.js [hunk 5]
```diff
@@ -21,40 +21,49 @@
...
+  if (goog.isDefAndNotNull(encHeader) && goog.isDefAndNotNull(
+      encHeader.value)) {
+    encoding = encHeader.value;
+  }
+
+  try {
+    filename = node.header[constants.Mime.CONTENT_DISPOSITION].params.filename;
+    encoding = node.header[constants.Mime.CONTENT_TRANSFER_ENCODING].value;
+  } catch (e) {
+    throw new e2e.openpgp.error.UnsupportedError('');
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/utils.js]: Add error messages here and below.

File: src/javascript/crypto/e2e/openpgp/pgpmime/utils.js [hunk 6]
```diff
@@ -111,15 +120,204 @@ e2e.openpgp.pgpmime.Utils.parseHeaderValue = function(text) {
...
+  });
+  // Implicit content-transfer-encoding is 7bit (RFC 2045).
+  goog.object.setIfUndefined(parsed, constants.Mime.CONTENT_TRANSFER_ENCODING, {
+    value: constants.Mime.SEVEN_BIT
+  });
+
+  var headerLines = utils.splitHeaders_(text, opt_separator);
+  goog.array.forEach(headerLines, function(line) {
+    // Ex: 'Content-Type: multipart/encrypted'
+    var parts = line.split(':');
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/utils.js]: Unless you can assure there will be no more colons, use goog.string.split with limit. That's also for all other uses of String.split below.

File: src/javascript/crypto/e2e/openpgp/pgpmime/utils.js [hunk 7]
```diff
@@ -111,15 +120,204 @@ e2e.openpgp.pgpmime.Utils.parseHeaderValue = function(text) {
...
+  if (goog.isDefAndNotNull(opt_separator)) {
+    separator = opt_separator;
+  }
+  var splitter = new RegExp(separator + '(?![\x20\x09]+)');
+  // Per RFC 2047 part 2., a multiline header should be wrapped by CRLF SPACE.
+  // Distinct header lines should be separated from each other by a sole CRLF.
+  // In practice, many email systems use an arbitrary amount of tabs and
+  // whitespaces instead of a single SPACE. To accomodate this, we separate
+  // headers by looking for CRLFs that are not followed by any variable amount
+  // of tabs or whitespaces.
```
Comment (koto) [inline src/javascript/crypto/e2e/openpgp/pgpmime/utils.js]: // of tabs or spaces.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js [hunk 8]
```diff
@@ -130,10 +125,66 @@ e2e.openpgp.pgpmime.MimeNode.prototype.setContent = function(content) {
...
+ * @param {string} boundary The proposed boundary.
+ * @return {boolean} Returns true if the proposed boundary does not appear
+ *     within the message. Otherwise returns false.
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.isBoundaryValid_ = function(boundary) {
+  // Assign a temporary boundary that will be used in buildMessage().
+  this.boundary_ = 'temporary_boundary';
+  var tempMessage = this.mimeToString_();
+  return !goog.string.contains(tempMessage, boundary);
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: For extra safety, check that the boundary is < 70 characters, as required by RFC 2046.

File: src/javascript/crypto/e2e/openpgp/pgpmime/constants.js (hunk 1 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/constants.js]: My list was incomplete, thanks for pointing this out. 
I'll push a complete list of image subtypes, base on this list:
http://www.iana.org/assignments/media-types/media-types.xhtml#image

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 2 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: I'll change it to a `do..while` loop here. We're relying on `goog.string.getRandomString()` - do you think that's sufficiently random?

File: src/javascript/crypto/e2e/openpgp/pgpmime/utils.js (hunk 6 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/utils.js]: Thanks for pointing that out - it seems like good practice. I'll add it wherever it's applicable.

File: src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js (hunk 8 above)
Comment (yonigoogle) [inline src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js]: Will do, thanks!

File: src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js [hunk 9]
```diff
@@ -119,14 +119,25 @@ e2e.openpgp.pgpmime.PgpMail.prototype.buildMimeTree = function() {
     textNode.setContent(this.content.body);
 
     goog.array.forEach(this.content.attachments, function(attachment) {
+      var contentType = constants.Mime.OCTET_STREAM;
+      // Implicit content-transfer-encoding is 7bit (RFC 2045).
+      var contentTransferEncoding = constants.Mime.SEVEN_BIT;
+
+      if (goog.isDefAndNotNull(attachment.type)) {
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/openpgp/pgpmime/pgpmail.js]: `type` is no longer defined on `e2e.openpgp.pgpmime.types.Attachment`

**Pull request merged**